Synced flush backport

brwe opened the pull request:
Was relatively unproblematic, only SyncedFlushService was a little hairy

spinscale commented on src/main/java/org/elasticsearch/indices/flush/SyncedFlushService.java:
```diff
@@ -358,7 +410,7 @@ public void onResponse(PreSyncedFlushResponse response) {
                 @Override
                 public void onFailure(Throwable e) {
-                    logger.trace("{} error while performing pre synced flush on [{}], skipping", e, shardId, shard);
+                    logger.trace("{} error while performing pre synced flush on [{}], skipping", shardId, e, shard);
```
this looks flipped and does not log the exception?

bleskes replied:
++

spinscale commented on src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java:
```diff
@@ -178,6 +178,8 @@ public void phase1(final SnapshotIndexCommit snapshot) throws ElasticsearchExcep
                 }
                 // we shortcut recovery here because we have nothing to copy. but we must still start the engine on the target.
                 // so we don't return here
+                logger.trace("[{}][{}] skipping [phase1] to {} - identical sync id [{}] found on both source and target", indexName, shardId,
```
missing `[]` around targetNode?

bleskes replied:
Disco Nodes toString adds []

spinscale replied:
thx for the clarification

bleskes commented on src/main/java/org/elasticsearch/indices/flush/SyncedFlushService.java:
```diff
@@ -252,41 +315,24 @@ void sendSyncRequests(final String syncId, final List<ShardRouting> shards, Clus
+                contDownAndSendResponseIfDone(syncId, shards, shardId, totalShards, listener, countDown, results);
                 continue;
             }
             assert node.version().onOrAfter(Version.V_1_6_0) : " node with version " + node.version() + " can not have a sync commit ID";
```
not part of this change but I just noticed it - should be " can not have an *expected* commit ID " (not sync)

bleskes commented on src/main/java/org/elasticsearch/indices/flush/SyncedFlushService.java:
```diff
@@ -252,41 +315,24 @@ void sendSyncRequests(final String syncId, final List<ShardRouting> shards, Clus
-                        return performSyncedFlush(new SyncedFlushRequest(shard.shardId(), syncId, expectedCommitId));
-                    }
-                }, currentShardListener);
+                SyncedFlushResponse existing = results.put(shard, performSyncedFlush(new SyncedFlushRequest(shard.shardId(), syncId, expectedCommitId)));
```
why did you remove the FLUSH executor?

brwe replied:
I thought it was not needed anymore but think now it blocks and might not even be in the right thread pool...will add again. thanks also @spinscale

bleskes commented on src/test/java/org/elasticsearch/gateway/local/RecoveryBackwardsCompatibilityTests.java:
```diff
@@ -77,7 +77,8 @@ public void testReusePeerRecovery() throws Exception {
         if (randomBoolean()) { // just make sure it doesn't break anything - we seal before we actually bump replicas
-            backwardsCluster().internalCluster().client().admin().indices().prepareSealIndices("test").get();
+            logger.info("--> trying to sync flush");
+            assertEquals(SyncedFlushUtil.attemptSyncedFlush(internalCluster(), "test").failedShards(), 0);
```
don't we need backwardsCluster().internalCluster() here? also, are we sure we have at least one local (client) node? o.w. there is not SyncedFlushtService instance..

brwe replied:
yes, should be backwardsCluster()... we always have current version nodes in the test unless annotated otherwise. is that what you mean?

bleskes replied:
Yeah, that is what I meant. I guess we should check and protect for having no current version nodes, no?

On Fri, May 29, 2015 at 5:44 PM, Britta Weber <notifications@github.com>
wrote:

>> @@ -77,7 +77,8 @@ public void testReusePeerRecovery() throws Exception {
>>          indexRandom(true, builders);
>>          ensureGreen();
>>          if (randomBoolean()) { // just make sure it doesn't break anything - we seal before we actually bump replicas
>> -            backwardsCluster().internalCluster().client().admin().indices().prepareSealIndices("test").get();
>> +            logger.info("--> trying to sync flush");
>> +            assertEquals(SyncedFlushUtil.attemptSyncedFlush(internalCluster(), "test").failedShards(), 0);
> yes, should be backwardsCluster()... we always have current version nodes in the test unless annotated otherwise. is that what you mean?
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/elastic/elasticsearch/pull/11417/files#r31336414

brwe replied:
I will add that, also to indexRandom() in ElasticsearchIntegrationTest. However, after running the test I found that the actual problem is that we check for no failures here - which is not good because the check is doomed to fail if a shard was an older version node. will remove the assert as well.

bleskes commented:
Thx @brwe for picking this up. Left some minor comments

spinscale commented:
did not spot any huge differences on my side, except the same question @bleskes had about removing the FLUSH executor...

apart from that LGTM

brwe commented:
ok, all done. I let the bwc tests run and found that there we also have to check if it is a bwc test or not when we call indexRandom(). pushed a fix for that.

brwe commented:
fixed the bwc test and added checks to make sure we actually have at least one current version node when we run the bwc tests. want to take another look?

bleskes commented on src/test/java/org/elasticsearch/test/ElasticsearchIntegrationTest.java:
```diff
@@ -1480,8 +1478,22 @@ private void postIndexAsyncActions(String[] indices, List<CountDownLatch> inFlig
+                        // bwc test
+                        if (cluster() instanceof CompositeTestCluster) {
+                            CompositeTestCluster compositeTestCluster = (CompositeTestCluster) cluster();
+                            if (compositeTestCluster.numNewDataNodes() > 0) {
```
can we add a comment as to why this is needed here? I'm not sure it's trivial to figure out.

bleskes commented:
LGTM. Left one little comment. Feel free to push.
